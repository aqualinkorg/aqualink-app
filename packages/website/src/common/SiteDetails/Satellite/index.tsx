import React from "react";
import {
  withStyles,
  WithStyles,
  createStyles,
  Card,
  CardContent,
  Typography,
  CardHeader,
  Grid,
  Tooltip,
  Box,
} from "@material-ui/core";

import { dhwColorCode } from "../../../assets/colorCode";
import type { LiveData } from "../../../store/Reefs/types";
import { formatNumber } from "../../../helpers/numberUtils";
import satellite from "../../../assets/satellite.svg";
import {
  dhwColorFinder,
  degreeHeatingWeeksCalculator,
} from "../../../helpers/degreeHeatingWeeks";
import { styles as incomingStyles } from "../styles";
import UpdateInfo from "../../UpdateInfo";
import { toRelativeTime } from "../../../helpers/dates";

const Satellite = ({ maxMonthlyMean, liveData, classes }: SatelliteProps) => {
  const { degreeHeatingDays, satelliteTemperature, sstAnomaly } = liveData;
  const relativeTime =
    satelliteTemperature?.timestamp &&
    toRelativeTime(satelliteTemperature.timestamp);

  const degreeHeatingWeeks = degreeHeatingWeeksCalculator(
    degreeHeatingDays?.value
  );

  const metrics = [
    {
      label: "SURFACE TEMP",
      value: `${formatNumber(satelliteTemperature?.value, 1)}°C`,
    },
    {
      label: "HISTORICAL MAX",
      value: `${formatNumber(maxMonthlyMean, 1)}°C`,
      tooltipTitle: "Historical maximum monthly average over the past 20 years",
    },
    {
      label: "DEGREE HEATING WEEKS",
      value: formatNumber(degreeHeatingWeeks, 1),
      tooltipTitle:
        "Degree Heating Weeks - a measure of the amount of time above the 20 year historical maximum temperatures",
    },
    {
      label: "SST ANOMALY",
      tooltipTitle: "Difference between current SST and longterm average",
      value: `${
        sstAnomaly
          ? `${sstAnomaly > 0 ? "+" : ""}${formatNumber(sstAnomaly, 1)}`
          : "- -"
      }°C`,
    },
  ];

  return (
    <Card
      className={classes.card}
      style={{ backgroundColor: dhwColorFinder(degreeHeatingWeeks) }}
    >
      <CardHeader
        className={classes.header}
        title={
          <Grid container>
            <Grid item>
              <Typography className={classes.cardTitle} variant="h6">
                SATELLITE OBSERVATION
              </Typography>
            </Grid>
          </Grid>
        }
      />

      <CardContent className={classes.content}>
        <Box p="1rem" display="flex" flexGrow={1}>
          <Grid container spacing={1}>
            {metrics.map(({ label, value, tooltipTitle }) => (
              <Grid key={label} item xs={6}>
                <Typography
                  className={classes.contentTextTitles}
                  variant="subtitle2"
                >
                  {label}
                </Typography>
                <Tooltip title={tooltipTitle || ""}>
                  <Typography
                    className={classes.contentTextValues}
                    variant="h3"
                  >
                    {value}
                  </Typography>
                </Tooltip>
              </Grid>
            ))}
          </Grid>
        </Box>

        <Grid container>
          {dhwColorCode.map(({ value, color }) => (
            <Grid
              key={value}
              item
              xs={1}
              style={{ backgroundColor: `${color}`, height: "2rem" }}
            >
              <Box textAlign="center">
                <Typography variant="caption" align="center">
                  {value}
                </Typography>
              </Box>
            </Grid>
          ))}
        </Grid>

        <UpdateInfo
          relativeTime={relativeTime}
          timeText="Last data received"
          image={satellite}
          imageText="NOAA"
          live={false}
          frequency="daily"
          href="https://coralreefwatch.noaa.gov/"
        />
      </CardContent>
    </Card>
  );
};

const styles = () =>
  createStyles({
    ...incomingStyles,
    card: {
      ...incomingStyles.card,
      display: "flex",
      flexDirection: "column",
      height: "100%",
    },
    content: {
      display: "flex",
      flexDirection: "column",
      justifyContent: "space-between",
      flexGrow: 1,
      padding: 0,
    },
    contentText: {
      marginTop: "1rem",
      padding: "0 1rem",
    },
  });

interface SatelliteIncomingProps {
  maxMonthlyMean: number | null;
  liveData: LiveData;
}

type SatelliteProps = WithStyles<typeof styles> & SatelliteIncomingProps;

export default withStyles(styles)(Satellite);
