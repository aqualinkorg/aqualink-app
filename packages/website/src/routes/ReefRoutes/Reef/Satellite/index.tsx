import React from "react";
import {
  withStyles,
  WithStyles,
  createStyles,
  Theme,
  Card,
  CardContent,
  Typography,
  CardHeader,
  Grid,
} from "@material-ui/core";

import { colorCode } from "../../../../assets/colorCode";
import type { Data } from "../../../../store/Reefs/types";
import { sortDailyData } from "../../../../helpers/sortDailyData";
import { formatNumber } from "../../../../helpers/numberUtils";
import satellite from "../../../../assets/satellite.svg";
import { degreeHeatingWeeksCalculator } from "../../../../helpers/degreeHeatingWeeks";

const Satellite = ({ dailyData, classes }: SatelliteProps) => {
  const sortByDate = sortDailyData(dailyData, "desc");
  const { degreeHeatingDays, satelliteTemperature } = sortByDate[0];

  const degreeHeatingWeeks = degreeHeatingWeeksCalculator(degreeHeatingDays);

  return (
    <Card className={classes.card}>
      <CardHeader
        className={classes.header}
        title={
          <Grid container alignItems="center" justify="space-between">
            <Grid item xs={8}>
              <Typography className={classes.cardTitle} variant="h6">
                SATELLITE OBSERVATION
              </Typography>
            </Grid>
            <Grid item xs={1}>
              <img alt="satellite" src={satellite} />
            </Grid>
          </Grid>
        }
      />
      <CardContent className={classes.content}>
        <Grid
          container
          direction="column"
          alignItems="stretch"
          justify="space-between"
          style={{ height: "100%" }}
        >
          <Grid className={classes.contentText} item>
            <Grid container spacing={3}>
              <Grid item xs={12}>
                <Typography
                  className={classes.contentTextTitles}
                  color="textPrimary"
                  variant="subtitle2"
                >
                  SURFACE TEMP
                </Typography>
                <Typography
                  className={classes.contentTextValues}
                  color="textPrimary"
                  variant="h2"
                >
                  {satelliteTemperature
                    ? `${formatNumber(satelliteTemperature, 1)} \u2103`
                    : "- -"}
                </Typography>
              </Grid>
              <Grid item xs={12}>
                <Typography
                  className={classes.contentTextTitles}
                  color="textPrimary"
                  variant="subtitle2"
                >
                  DEGREE HEATING WEEKS
                </Typography>
                <Typography
                  className={classes.contentTextValues}
                  color="textPrimary"
                  variant="h2"
                >
                  {formatNumber(degreeHeatingWeeks, 1) || "- -"}
                </Typography>
              </Grid>
            </Grid>
          </Grid>
          <Grid item container>
            {colorCode.map((elem) => (
              <Grid
                container
                justify="center"
                alignItems="center"
                key={elem.value}
                item
                xs={1}
              >
                <Grid
                  container
                  justify="center"
                  alignItems="center"
                  item
                  style={{ backgroundColor: `${elem.color}`, height: "2rem" }}
                >
                  <Typography variant="caption">{elem.value}</Typography>
                </Grid>
              </Grid>
            ))}
          </Grid>
        </Grid>
      </CardContent>
    </Card>
  );
};

const styles = (theme: Theme) =>
  createStyles({
    card: {
      height: "100%",
      width: "100%",
      backgroundColor: "#0c9da5",
      display: "flex",
      flexDirection: "column",
    },
    cardTitle: {
      lineHeight: 1.5,
      margin: "0 0 0.5rem 1rem",
    },
    header: {
      flex: "0 1 auto",
      padding: "0.5rem 1.5rem 0 1rem",
    },
    content: {
      flex: "1 1 auto",
      padding: 0,
    },
    contentText: {
      padding: "2rem 3rem 0 3rem",
    },
    contentTextTitles: {
      lineHeight: 1.33,
    },
    contentTextValues: {
      fontWeight: 300,
      [theme.breakpoints.between("md", "lg")]: {
        fontSize: 32,
      },
    },
  });

interface SatelliteIncomingProps {
  dailyData: Data[];
}

type SatelliteProps = WithStyles<typeof styles> & SatelliteIncomingProps;

export default withStyles(styles)(Satellite);
