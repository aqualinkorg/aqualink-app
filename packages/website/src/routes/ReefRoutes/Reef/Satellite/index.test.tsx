import React from "react";
import { render } from "@testing-library/react";

import Satellite from ".";

test("renders as expected", () => {
  const liveData = {
    reef: { id: 1 },
    date: "2020-07-01T14:25:18.008Z",
    bottomTemperature: {
      value: 25,
      timestamp: "2020-07-01T14:25:18.008Z",
    },
    satelliteTemperature: {
      value: 26,
      timestamp: "2020-07-01T14:25:18.008Z",
    },
    degreeHeatingDays: {
      value: 32,
      timestamp: "2020-07-01T14:25:18.008Z",
    },
  };
  const { container } = render(
    <Satellite maxMonthlyMean={24} liveData={liveData} />
  );
  expect(container).toMatchSnapshot();
});
