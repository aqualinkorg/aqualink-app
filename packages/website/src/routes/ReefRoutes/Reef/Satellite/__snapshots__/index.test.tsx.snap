// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <mock-card
    classname="Satellite-card-7"
    style="background-color: rgb(117, 180, 114);"
  >
    <div
      class="MuiCardHeader-root Satellite-header-2"
    >
      <div
        class="MuiCardHeader-content"
      >
        <mock-typography
          classname="MuiCardHeader-title"
          component="span"
          display="block"
          variant="h5"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-space-between"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-8"
            >
              <mock-typography
                classname="Satellite-cardTitle-1"
                variant="h6"
              >
                SATELLITE OBSERVATION
              </mock-typography>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
            >
              <img
                alt="satellite"
                class="Satellite-titleImage-8"
                src="satellite.svg"
              />
            </div>
          </div>
        </mock-typography>
      </div>
    </div>
    <mock-cardcontent
      classname="Satellite-content-9"
    >
      <mock-box
        display="flex"
        flexgrow="1"
        p="1rem"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
          >
            <mock-typography
              classname="Satellite-contentTextTitles-3"
              variant="subtitle2"
            >
              SURFACE TEMP
            </mock-typography>
            <mock-typography
              classname="Satellite-contentTextValues-4"
              variant="h2"
            >
              26.0 °C
            </mock-typography>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
          >
            <mock-typography
              classname="Satellite-contentTextTitles-3"
              variant="subtitle2"
            >
              HISTORICAL MAX
            </mock-typography>
            <mock-tooltip
              title="Historical maximum monthly average over the past 20 years"
            >
              <mock-typography
                classname="Satellite-contentTextValues-4"
                variant="h2"
              >
                24.0 °C
              </mock-typography>
            </mock-tooltip>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <mock-typography
              classname="Satellite-contentTextTitles-3"
              variant="subtitle2"
            >
              HEAT STRESS
            </mock-typography>
            <mock-tooltip
              title="Degree Heating Weeks - a measure of the amount of time above the 20 year historical maximum temperatures"
            >
              <mock-typography
                classname="Satellite-contentTextValues-4"
                variant="h2"
              >
                4.6 DHW
              </mock-typography>
            </mock-tooltip>
          </div>
        </div>
      </mock-box>
      <div
        class="MuiGrid-root MuiGrid-container"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
          style="background-color: rgb(8, 91, 163); height: 2rem;"
        >
          <mock-box
            textalign="center"
          >
            <mock-typography
              align="center"
              variant="caption"
            >
              0
            </mock-typography>
          </mock-box>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
          style="background-color: rgb(12, 157, 165); height: 2rem;"
        >
          <mock-box
            textalign="center"
          >
            <mock-typography
              align="center"
              variant="caption"
            >
              2
            </mock-typography>
          </mock-box>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
          style="background-color: rgb(117, 180, 114); height: 2rem;"
        >
          <mock-box
            textalign="center"
          >
            <mock-typography
              align="center"
              variant="caption"
            >
              4
            </mock-typography>
          </mock-box>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
          style="background-color: rgb(123, 217, 37); height: 2rem;"
        >
          <mock-box
            textalign="center"
          >
            <mock-typography
              align="center"
              variant="caption"
            >
              6
            </mock-typography>
          </mock-box>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
          style="background-color: rgb(193, 229, 34); height: 2rem;"
        >
          <mock-box
            textalign="center"
          >
            <mock-typography
              align="center"
              variant="caption"
            >
              8
            </mock-typography>
          </mock-box>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
          style="background-color: rgb(229, 187, 43); height: 2rem;"
        >
          <mock-box
            textalign="center"
          >
            <mock-typography
              align="center"
              variant="caption"
            >
              10
            </mock-typography>
          </mock-box>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
          style="background-color: rgb(230, 121, 40); height: 2rem;"
        >
          <mock-box
            textalign="center"
          >
            <mock-typography
              align="center"
              variant="caption"
            >
              12
            </mock-typography>
          </mock-box>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
          style="background-color: rgb(232, 79, 46); height: 2rem;"
        >
          <mock-box
            textalign="center"
          >
            <mock-typography
              align="center"
              variant="caption"
            >
              14
            </mock-typography>
          </mock-box>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
          style="background-color: rgb(216, 69, 36); height: 2rem;"
        >
          <mock-box
            textalign="center"
          >
            <mock-typography
              align="center"
              variant="caption"
            >
              16
            </mock-typography>
          </mock-box>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
          style="background-color: rgb(128, 58, 32); height: 2rem;"
        >
          <mock-box
            textalign="center"
          >
            <mock-typography
              align="center"
              variant="caption"
            >
              18
            </mock-typography>
          </mock-box>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
          style="background-color: rgb(185, 41, 68); height: 2rem;"
        >
          <mock-box
            textalign="center"
          >
            <mock-typography
              align="center"
              variant="caption"
            >
              20
            </mock-typography>
          </mock-box>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
          style="background-color: rgb(128, 48, 126); height: 2rem;"
        >
          <mock-box
            textalign="center"
          >
            <mock-typography
              align="center"
              variant="caption"
            >
              25
            </mock-typography>
          </mock-box>
        </div>
      </div>
    </mock-cardcontent>
  </mock-card>
</div>
`;
